storage: regression test leaked intents on bounced proposal

tbg opened the pull request:
This adds the test promised in the PR below. When a transaction
committed but the commit applied at an invalid lease applied index,
we'd formerly (due to a recent change) leak the intents as committed
which would cause dirty writes. Adapt an existing test to roughly
do the following to prevent regression.

The test (now) sets up two ranges and lets a transaction (anchored on
the left) write to both of them. It then starts readers for both keys
written by the txn and waits for them to enter the txn wait queue. Next,
it lets the txn attempt to commit but injects a forced error below Raft.
The bugs would formerly notify the txn wait queue that the transaction
had committed (not true) and that its external intent (i.e. the one on
the right range) could be resolved (not true). Verify that neither
occurs.

See #34659.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34733)
<!-- Reviewable:end -->

petermattis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34733)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/kv/integration_test.go, line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-LYNlKHq3iw_QECwQTY2:-LYNlKHq3iw_QECwQTY3:bcq0c63) ([raw file](https://github.com/cockroachdb/cockroach/blob/72a1ed37e5d27b822a2d5b2042cb14b912ade283/pkg/kv/integration_test.go#L197)):*
> ```Go
> // transaction as Aborted. This test is a regression test for #30792 which was
> // causing pushers in the txn wait queue to consider such a transaction
> // committed. It is also a regression test for a the similar bug [1] in which
> ```

`s/a the/a/g`
___
*[pkg/kv/integration_test.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-LYNmO_71BlX5gyHuPOP:-LYNmO_71BlX5gyHuPOQ:b-tb395e) ([raw file](https://github.com/cockroachdb/cockroach/blob/72a1ed37e5d27b822a2d5b2042cb14b912ade283/pkg/kv/integration_test.go#L263)):*
> ```Go
> 						if illegalLeaseIndex {
> 							illegalLeaseIndex = false
> 							return 1, roachpb.NewErrorf("test injected err (illegal lease index)")
> ```

Add a comment that `1` corresponds to `proposalIllegalLeaseIndex`. Or copy that constant here (with a comment that we don't want the dependency cycle of referring to the same value).
___
*[pkg/kv/integration_test.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-LYNm08-ATd2Jx0m0uxs:-LYNm08-ATd2Jx0m0uxt:b-sq70ae) ([raw file](https://github.com/cockroachdb/cockroach/blob/72a1ed37e5d27b822a2d5b2042cb14b912ade283/pkg/kv/integration_test.go#L327)):*
> ```Go
> 				readerDone <- fmt.Errorf("%s: expected value to not exist, got: %s", key, val)
> 			}
> 			readerDone <- nil
> ```

This method looks like is could send on `readerDone` more than once. That doesn't seem right.
___
*[pkg/kv/integration_test.go, line 331 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-LYNlcJlDq3VfJya3LXw:-LYNlcJlDq3VfJya3LXx:b-s654a9) ([raw file](https://github.com/cockroachdb/cockroach/blob/72a1ed37e5d27b822a2d5b2042cb14b912ade283/pkg/kv/integration_test.go#L331)):*
> ```Go
> 	}
> 
> 	// Wait for the reader to enter the txn wait queue.
> ```

`Wait for both reads to enter the txn wait queue.`


<!-- Sent from Reviewable.io -->

tbg committed:
storage: regression test leaked intents on bounced proposal

This adds the test promised in the PR below. When a transaction
committed but the commit applied at an invalid lease applied index,
we'd formerly (due to a recent change) leak the intents as committed
which would cause dirty writes. Adapt an existing test to roughly
do the following to prevent regression.

The test (now) sets up two ranges and lets a transaction (anchored on
the left) write to both of them. It then starts readers for both keys
written by the txn and waits for them to enter the txn wait queue. Next,
it lets the txn attempt to commit but injects a forced error below Raft.
The bugs would formerly notify the txn wait queue that the transaction
had committed (not true) and that its external intent (i.e. the one on
the right range) could be resolved (not true). Verify that neither
occurs.

See #34659.

Release note: None

tbg reviewed (commented):
TFTR, PTAL!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-:-LYRK3dq7mtEfmunsZd_:b-qs6e2f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @petermattis)

___
*[pkg/kv/integration_test.go, line 197 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-LYNlKHq3iw_QECwQTY2:-LYRIph17ObT2V2H5jFa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/72a1ed37e5d27b822a2d5b2042cb14b912ade283/pkg/kv/integration_test.go#L197)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`s/a the/a/g`
</blockquote></details>

Done.
___
*[pkg/kv/integration_test.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-LYNmO_71BlX5gyHuPOP:-LYRJUN9DSFGYpTy3z2M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/72a1ed37e5d27b822a2d5b2042cb14b912ade283/pkg/kv/integration_test.go#L263)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Add a comment that `1` corresponds to `proposalIllegalLeaseIndex`. Or copy that constant here (with a comment that we don't want the dependency cycle of referring to the same value).
</blockquote></details>

Done.
___
*[pkg/kv/integration_test.go, line 327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-LYNm08-ATd2Jx0m0uxs:-LYRJmpJ3ECQQf109QwD:b-4c5ezp) ([raw file](https://github.com/cockroachdb/cockroach/blob/72a1ed37e5d27b822a2d5b2042cb14b912ade283/pkg/kv/integration_test.go#L327)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This method looks like is could send on `readerDone` more than once. That doesn't seem right.
</blockquote></details>

Good catch, done.
___
*[pkg/kv/integration_test.go, line 331 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-LYNlcJlDq3VfJya3LXw:-LYRJpFH6oiMJBgU2BL9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/72a1ed37e5d27b822a2d5b2042cb14b912ade283/pkg/kv/integration_test.go#L331)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`Wait for both reads to enter the txn wait queue.`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34733#-:-LYSbYYn8sKmCLhxFPxZ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg commented:
bors r=petermattis
